File: 1417_reformat/reformat.go
/*
 * @Date: 2022-08-11
 * @LastEditors: mengsen_wang@163.com
 * @LastEditTime: 2022-08-11
 * @FilePath: /algorithm/1417_reformat/reformat.go
 */

package main

import "unicode"

func reformat(s string) string {
	sumDigit := 0
	for _, c := range s {
		if unicode.IsDigit(c) {
			sumDigit++
		}
	}
	sumAlpha := len(s) - sumDigit
	if func(x int) int {
		if x < 0 {
			return -x
		}
		return x
	}(sumDigit-sumAlpha) > 1 {
		return ""
	}
	flag := sumDigit > sumAlpha
	t := []byte(s)
	for i, j := 0, 1; i < len(t); i += 2 {
		if unicode.IsDigit(rune(t[i])) != flag {
			for unicode.IsDigit(rune(t[j])) != flag {
				j += 2
			}
			t[i], t[j] = t[j], t[i]
		}
	}
	return string(t)
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}
	assert(reformat("a0b1c2") == "a0b1c2")
	assert(reformat("leetcode") == "")
	assert(reformat("1229857369") == "")
	assert(reformat("covid2019") == "c2o0v1i9d")
	assert(reformat("ab123") == "1a2b3")
}

File: cpp/1417_reformat/reformat.cpp
/*
 * @Date: 2022-08-11
 * @LastEditors: mengsen_wang@163.com
 * @LastEditTime: 2022-08-11
 * @FilePath: /algorithm/1417_reformat/reformat.cpp
 */

#include <cassert>
#include <string>

using namespace std;

class Solution {
 public:
  string reformat(string s) {
    int sum_digit = 0;
    for (auto& c : s) {
      if (isdigit(c)) {
        sum_digit++;
      }
    }
    int sum_alpha = s.size() - sum_digit;
    if (abs(sum_digit - sum_alpha) > 1) {
      return "";
    }
    bool flag = sum_digit > sum_alpha;
    for (int i = 0, j = 1; i < s.size(); i += 2) {
      if (isdigit(s[i]) != flag) {
        while (isdigit(s[j]) != flag) {
          j += 2;
        }
        swap(s[i], s[j]);
      }
    }
    return s;
  }
};

int main() {
  assert(Solution().reformat("a0b1c2") == "0a1b2c");
  assert(Solution().reformat("leetcode") == "");
  assert(Solution().reformat("1229857369") == "");
  assert(Solution().reformat("covid2019") == "c2o0v1i9d");
  assert(Solution().reformat("ab123") == "1a2b3");
}